corehq/apps/api/es.py

```diff
@@ -701,7 +648,59 @@ def consume_params(self, raw_params):
 ]
 
 
-def es_search(request, domain, reserved_query_params=None):
+def _validate_and_get_es_filter(search_param):
+    #
+    try:
+        _filter = search_param['filter']
+        # custom use case by 'enveritas' project for Form API
+        date_range = _filter['range']['inserted_at']
+        return {
+            'range': {'inserted_at': date_range}
+        }
+    except KeyError:
+        pass
+    try:
+        # Data export tool passes below structure. validate that it is so
+        try:
+            _range = _filter['or'][0]['and'][0]['range']['server_modified_on']
+        except KeyError:
+            _range = _filter['or'][0]['and'][1]['range']['server_modified_on']
+        server_modified_missing = {"missing": {
+            "field": "server_modified_on", "null_value": True, "existence": True}
+        }
+        expected = json.dumps({
+            "or": [
+                {
+                    "and": [
+                        {
+                            "not": server_modified_missing
+                        },
+                        {
+                            "range": {
+                                "server_modified_on": _range
+                            }
+                        }
+                    ]
+                },
+                {
+                    "and": [
+                        server_modified_missing,
+                        {
+                            "range": {
+                                "received_on": _range
+                            }
+                        }
+                    ]
+                }
+            ]
+        }, sort_keys=True)
+        assert json.dumps(_filter, sort_keys=True) == expected
```

Comment:
this check doesn't feel right. what's the plan with these in the future, do the filters need to be changed for ES v7?

Reply:
Yeah, it's not a great check, for example it fails if the order of the lists inside is different (I tested these against the two known uses from logs which this check works well for). There is jsonschema that we could use, but I thought it's okay to do manually since it's just two inputs only. Do you have a better idea?

The filters will continue to be as they are since they come from outside of HQ (Data export tool and an API custom use case). For ES7, I will just transform them to ES7 filters.

Reply:
Here are my thoughts:
1. once we've re-indexed the form index we won't need to do this funny query any more since all data will have a `server_modified_on` populated. This means that whatever we do will be a short term solution.
2. Instead of forcing clients to deal with the inconsistent data in the index we should have HQ handle it so that if a client wants to filter by `server_modified_on` then we munge that to account for the potentially missing field in old forms.
3. If we do *2 then all we need to do is detect whether or not the incoming request is trying to filter by `server_modified_on` and get the range (instead of trying to match the entire query).

```python

class XFormServerModifiedParams:
    param = 'server_modified_on'

    def consume_params(self, raw_params):
        value = raw_params.pop(self.param, None)
        if value:
            return filters.OR([
                filters.AND([
                    filters.NOT(filters.missing(self.param)), filters.term(self.param, value)
                ]),
                filters.AND([
                    filters.missing(self.param), filters.term("received_on", value)
                ])
            ])


_range = None
try:
    _range = _filter['or'][0]['and'][0]['range']['server_modified_on']
except KeyError:
    try:
        _range = _filter['or'][0]['and'][1]['range']['server_modified_on']
    except KeyError:
        pass

if _range:
    return XFormServerModifiedParams().consume_params({'server_modified_on': _range)
```

Reply:
https://github.com/dimagi/commcare-hq/pull/28032/commits/8da0c623987201beb0cf66be2c0fcbcfa93d247d